serialization.md

```diff
@@ -464,66 +458,93 @@ Merkle Patricia Tree
+    rule MerklePut ( MerkleLeaf ( LEAFPATH, LEAFVALUE ), PATH, VALUE )
       => #merkleExtensionBuilder( .ByteArray, LEAFPATH, LEAFVALUE, PATH, VALUE )
       requires #unparseByteStack( LEAFPATH ) =/=String #unparseByteStack( PATH )
        andBool #sizeByteArray( LEAFPATH ) >Int 0
        andBool #sizeByteArray( PATH )     >Int 0
        andBool LEAFPATH[0] ==Int PATH[0]
 
-    rule MerkleUpdate ( MerkleExtension ( EXTPATH, EXTTREE ), PATH, VALUE )
-      => MerkleExtension ( EXTPATH, MerkleUpdate ( EXTTREE, .ByteArray, VALUE ) )
+    rule MerklePut ( MerkleExtension ( EXTPATH, EXTTREE ), PATH, VALUE )
+      => MerkleExtension ( EXTPATH, MerklePut ( EXTTREE, .ByteArray, VALUE ) )
       requires #unparseByteStack( EXTPATH ) ==String #unparseByteStack( PATH )
 
-    rule MerkleUpdate ( MerkleExtension ( EXTPATH, EXTTREE ), PATH, VALUE )
-      => #merkleExtensionBrancher( MerkleUpdate( .MerkleBranch, PATH, VALUE ), EXTPATH, EXTTREE )
+    rule MerklePut ( MerkleExtension ( EXTPATH, EXTTREE ), PATH, VALUE )
+      => #merkleExtensionBrancher( MerklePut( MerkleBranch( .Map, "" ), PATH, VALUE ), EXTPATH, EXTTREE )
       requires #sizeByteArray( PATH ) >Int 0
        andBool EXTPATH[0] =/=Int PATH[0]
 
-    rule MerkleUpdate ( MerkleExtension ( EXTPATH, EXTTREE ), PATH, VALUE )
+    rule MerklePut ( MerkleExtension ( EXTPATH, EXTTREE ), PATH, VALUE )
       => #merkleExtensionSplitter( .ByteArray, EXTPATH, EXTTREE, PATH, VALUE )
       requires #unparseByteStack( EXTPATH ) =/=String #unparseByteStack( PATH )
        andBool #sizeByteArray( PATH ) >Int 0
        andBool EXTPATH[0] ==Int PATH[0]
 
-    rule MerkleUpdate ( MerkleBranch( M, _ ), PATH, VALUE )
+    rule MerklePut ( MerkleBranch( M, _ ), PATH, VALUE )
       => MerkleBranch( M, VALUE )
       requires #sizeByteArray( PATH ) ==Int 0
 
-    rule MerkleUpdate ( MerkleBranch( M, BRANCHVALUE ), PATH, VALUE )
+    rule MerklePut ( MerkleBranch( M, BRANCHVALUE ), PATH, VALUE )
       => #merkleUpdateBranch ( M, BRANCHVALUE, PATH[0], PATH[1 .. #sizeByteArray(PATH) -Int 1], VALUE )
       requires #sizeByteArray( PATH ) >Int 0
+
+    rule MerkleDelete( .MerkleTree, _ ) => .MerkleTree
+
+    rule MerkleDelete( MerkleLeaf( LPATH, V ), PATH ) => .MerkleTree                           requires #unparseByteStack(LPATH)  ==String #unparseByteStack(PATH)
+    rule MerkleDelete( MerkleLeaf( LPATH, V ), PATH ) => MerkleCheck( MerkleLeaf( LPATH, V ) ) requires #unparseByteStack(LPATH) =/=String #unparseByteStack(PATH)
+
+    rule MerkleDelete( MerkleExtension( EXTPATH, TREE ), PATH ) => MerkleExtension( EXTPATH, TREE ) requires #prefixLen(EXTPATH, PATH) =/=Int #sizeByteArray(EXTPATH)
+    rule MerkleDelete( MerkleExtension( EXTPATH, TREE ), PATH )
+      => MerkleCheck( MerkleExtension( EXTPATH, MerkleDelete( TREE, PATH[#sizeByteArray(EXTPATH) .. #sizeByteArray(PATH) -Int #sizeByteArray(EXTPATH)] ) ) )
+      requires #prefixLen(EXTPATH, PATH) ==Int #sizeByteArray(EXTPATH)
```

Comment:
This is the only place I see `#prefixLen` used, and it seems that we can make this simpler (and faster) by just defining a non-recursive function:

```
syntax Bool ::= #isPrefixByteArray ( ByteArray , ByteArray ) [function]
rule #isPrefixByteArray ( BA1 , BA2 ) => BA2 [ 0 .. #sizeByteArray(BA1) ] ==K BA1 requires #sizeByteArray(BA1) <=Int #sizeByteArray(BA2)
rule #isPrefixByteArray ( BA1 , BA2 ) => false [owise]
```

On the `Bytes` backends (ie LLVM and Haskell), `_[_.._]` will go to `substrBytes`, which should be very fast in comparison to the current recursive implementation of `#prefixLen`.